process-engine/process_engine_runtime
# ✨ Get instances and tasks by identity

**Requires:**
https://github.com/essential-projects/iam_contracts/pull/7 🔀 
https://github.com/process-engine/process_engine_contracts/pull/94 🔀 
https://github.com/process-engine/consumer_api_contracts/pull/50 🔀
https://github.com/process-engine/management_api_contracts/pull/36 🔀
https://github.com/process-engine/process_engine_core/pull/227 🔀
https://github.com/process-engine/flow_node_instance.repository.sequelize/pull/30 🔀 
https://github.com/process-engine/consumer_api_core/pull/79 🔀
https://github.com/process-engine/consumer_api_client/pull/27 🔀
https://github.com/process-engine/consumer_api_http/pull/22 🔀
https://github.com/process-engine/management_api_http/pull/28 🔀
https://github.com/process-engine/management_api_core/pull/34 🔀
https://github.com/process-engine/management_api_client/pull/29 🔀

**Changes:**

1. Fix the basic `start` UseCase on the `ExecuteProcessService`
    - The function now resolves immediately after starting the ProcessInstance
    - The functions return value now contains all relevant infos about that ProcessInstances, wrapped in a `ProcessStartedMessage` object
2. Refactor `EventAggregatorSettings`:
    - Merge `routePaths` and `messagePaths` into `messagePaths`
    - Rename `routeParams` to `messageParams`
    - Split messages into `internal` and `public` message types to differentiate between messages passed inside the process engine and messages that get send to a public subscriber.
    - Public messages will not contain the identity of the ProcessInstanceOwner, but the internal messages will.
4. Add `processInstanceOwner` to datamodels for `ManualTask` and `UserTask`
3. Add `owner` and `parentProcessInstanceId` to the `FlowNodeInstance` type
4. Fix the `onError` state transition for `CallActivities`, `Subprocesses` and internal `ServiceTasks`.
5. Implement the following Consumer API UseCases:
    - Request a users active ProcessInstances
    - Request a users waiting ManualTasks
    - Request a users waiting UserTasks
    - Subscribe for identity-specific notifications about a new UserTask
    - Subscribe for identity-specific notifications about a finished UserTask
    - Subscribe for identity-specific notifications about a new ManualTask
    - Subscribe for identity-specific notifications about a finished ManualTask
6. Add integrationtests for retrieving waiting UserTasks, ManualTasks and active ProcessInstances for a given identity
7. Add integrationtests for identity based notifications
8. Fix some cleanup issues with the existing ManualTask and UserTask tests.
9. Use `jsonwebtoken` to encode/decode all fixture identity tokens

**Issues:**

Closes #198 

PR: #200

## How can others test the changes?

- Request your own waiting UserTasks, ManualTasks or ProcessInstances through the new ConsumerAPI functions.
- Subscribe to the identity based notifications through the ConsumerAPI
- Run the tests and see that everything else still works as before.
- Afterwards, check the FlowNodeInstance test database for orphaned FlowNodeInstances (i.e. FlowNodes that never left the `running` or `suspended` state. There should no longer be any).

## PR-Checklist

Please check the boxes in this list after submitting your PR:
